ambry-clustermap/src/main/java/com.github.ambry.clustermap/ClusterMapManager.java

```diff
@@ -261,45 +283,149 @@ private boolean checkEnoughUnallocatedRawCapacity(int replicaCountPerDatacenter,
+   * NOTE 1: This method will change the ordering of the nodes in {@code dataNodes}
+   * NOTE 2: If {@code numDisks} valid disks could not be found, the returned list could be shorter than
+   *         {@code numDisks}
+   *
+   * @param dataNodes the list of {@link DataNode}s to sample from
+   * @param dataNodesUsed the set of {@link DataNode}s to exclude from the sample
+   * @param replicaCapacityInBytes the minimum amount of free space that a disk in the sample should have
+   * @param rackAware if {@code true}, only return disks in nodes that do not share racks with the nodes
+   *                  in {@code dataNodesUsed}
+   * @param numDisks how many disks to return in the sample
+   * @return a list of {@link Disk}s of length {@code numDisks}, or fewer if {@code numDisks} valid samples could not
+   *         be found
+   */
+  private List<Disk> getRandomDiskCandidateSample(List<DataNode> dataNodes, Set<DataNode> dataNodesUsed,
+      long replicaCapacityInBytes, boolean rackAware, int numDisks) {
+    Set<Long> rackIdsUsed = new HashSet<>();
+    if (rackAware) {
+      for (DataNode dataNode : dataNodesUsed) {
+        rackIdsUsed.add(dataNode.getRackId());
+      }
+    }
+    List<Disk> diskCandidates = new ArrayList<>();
+    int selectionBound = dataNodes.size();
+    Random randomGen = new Random();
+    while ((selectionBound > 0) && (diskCandidates.size() < numDisks)) {
+      int selectionIndex = randomGen.nextInt(selectionBound);
+      DataNode candidate = dataNodes.get(selectionIndex);
+      if (!dataNodesUsed.contains(candidate) && !rackIdsUsed.contains(candidate.getRackId())) {
+        Disk diskWithMostCapacity = getDiskWithMostUnallocatedRawCapacity(candidate, replicaCapacityInBytes);
+        if (diskWithMostCapacity != null) {
+          diskCandidates.add(diskWithMostCapacity);
+        }
+      }
+      selectionBound--;
+      Collections.swap(dataNodes, selectionIndex, selectionBound);
+    }
+    return diskCandidates;
+  }
+
+  /**
+   * Return a list of disks for a new partition in the specified {@link Datacenter}.  Does not retry if fewer than
+   * {@code replicaCountPerDatacenter} disks cannot be allocated.
+   *
+   * @param replicaCountPerDatacenter how many replicas to attempt to allocate in the datacenter
+   * @param replicaCapacityInBytes the minimum amount of free space on a disk for a replica
+   * @param datacenter the {@link Datacenter} to allocate replicas in
+   * @param disableRackAwareness if {@code true}, do not attempt a rack-aware allocation
+   * @return
+   */
   private List<Disk> allocateDisksForPartition(int replicaCountPerDatacenter, long replicaCapacityInBytes,
-      List<Datacenter> dataCenters) {
-    ArrayList<Disk> allocatedDisks = new ArrayList<Disk>();
+      Datacenter datacenter, boolean disableRackAwareness) {
+    ArrayList<Disk> disksToAllocate = new ArrayList<Disk>();
+    boolean rackAware = datacenter.isRackAware() && !disableRackAwareness;
```

Comment:
seems to me like we should throw here if `disableRackAwareness` is `false` but `dataCenter.isRackAware()` is `false`. Otherwise doesn't it mean you are silently no respecting one of the args provided?